### ISSUE TITLE
[Feature Request] Sequence Parallel to improve compute strength and save HBM space when LLM training. 

### ISSUE THREAD

**Issue by MoFHeka** [openxla/xla#8210]
Feature sequence parallel has already support by Megatron and DeepSpeed for a long time.
Input samples are not only separated at batch dim, but also at sequence length dim when using tensor parallelism.
Paper:https://arxiv.org/pdf/2105.13120.pdf
<img width="1149" alt="截屏2024-01-05 19 53 27" src="https://github.com/openxla/xla/assets/90189118/52950938-7247-4e9d-985f-f8a5ac11ef82">

Please:

- [x] Check for duplicate requests.
- [x] Describe your goal, and if possible provide a code snippet with a motivating example.

**Comment by cheshire** [comment]
Could you expand on the request? Is this an XLA-side optimization, or is it something which should be done in the modeling code?

**Comment by MoFHeka** [comment]
@Cheshire In Torch, people usually do it by the modeling code. 
But for XLA distributed training, it's hard to separate the input tensor at both batch and sequence dim by limitation of SPMD. 
For now I can't find a practicable way to do it by using jax.vmap and jax.pmap.

**Comment by cheshire** [comment]
@hawkinsp @jurahul WDYT?

**Comment by MoFHeka** [comment]
@jurahul Hello~? Any idea?

**Comment by MoFHeka** [comment]
@cheshire @jurahul https://github.com/NVIDIA/TransformerEngine/pull/602
It seems that Transformer Engine from Nvidia has implemented sequence parallelism. Is there anything we can learn to complete auto_sharding?

**Comment by dathudeptrai** [comment]
@MoFHeka alpha is all you need https://arxiv.org/pdf/2201.12023.pdf. https://github.com/alpa-projects/alpa. It should be already better than deepspeed.

**Comment by MoFHeka** [comment]
@dathudeptrai I chat with Alpa guys last years, At that time they had stopped maintaining Alpa. Hao Zhang suggested me to use Megatron instead, which performance is better than Alpa. Also he said there is no future if I do some effort in compiler optimization filed for LLM training. It suggested that if we don't use Google TPU, giving up Jax would be better.

**Comment by dathudeptrai** [comment]
@dathudeptrai Things might be different nowadays. Jax now already support Flash-attention, FP8 training that flawlessly integrated with the support from XLA. One problem with pytorch for example is that you have to custom the code quite a lot to support advanced training techniques. Also now you can call jax function in TF program which is great as well. 

About inference, I bet XLA still one of the best out-of-the-box optimization. For text generation, I found that XLA is 20 times faster than pytorch out-of-the-box. Even  pytorch + torch.compile + Flash-attention-2 + Group-query-attention  + KV-quantization still can not beat XLA out-of-the-box.

**Comment by MoFHeka** [comment]
@dathudeptrai I don't doubt that Jax already has many capabilities in the field of LLM training that it didn't have before. This is very good, and I also think implementation of Jax is very elegant. But I will not easily try Jax LLM training without a clear R&D roadmap and with risking a lack of project maintenance. After all, our team is responsible for the efficient computing of hundreds of A800/H800 servers. We don’t want to waste too much time for solving frameworks bugs and compatibility problems. That's why I wanted to ask Jax and the XLA team if they have plans to integrate and maintain the current various optimization methods into the main branch.

**Comment by dathudeptrai** [comment]
@MoFHeka I built my whole pipeline with Pytorch + Deepspeed/Megatron for training. For inference, I used whatever the fastest :). Normally the fastest one in the inference will be Jax/TF. Now I want to go beyond PyTorch and for sure TF lacks so many training techniques so I looked at Jax and now start experimenting with jax for some components in training. I think jax and xla team is working on pallas (the one used for implementing flash-atention in Jax). 

BTW, I am heavily focus on the recent opensource https://github.com/google/maxtext.

**Comment by MoFHeka** [comment]
@dathudeptrai  I noticed MaxText for several months. But it seems to have no interest in supporting GPU LLM training, which is unfortunate. I don't know if Google or any company has plans to continue develop and maintain Jax LLM training framework, at least for now it seems that no one at Alpa has taken over. First, at least integrate FP8 training and sequence parallelism into framework.

Besides, we also use Pytorch + Deepspeed/Megatron for training. And maybe mixed Ray pipeline for RLHF and multi-modal model. TensorRT-LLM has the strongest online inference throughput with comprehensive ability in our scenario. It's better than other frameworks, such as vllm.

**Comment by dathudeptrai** [comment]
@MoFHeka TensorRT-LLM still slower than XLA if you use XLA correctly :). 

FP8 training is already support in jax and TF as well :). MAxtext can be applied for GPU not only TPU. Flash-attention kernel in jax provided for both  GPU and TPU. I don't think the problem with GPU is actually a problem with jax. I do know that TPU has some optimization but it still covers mostly in GPU.

**Comment by MoFHeka** [comment]
> @MoFHeka TensorRT-LLM still slower than XLA if you use XLA correctly :).
> 
> FP8 training is already support in jax and TF as well :). MAxtext can be applied for GPU not only TPU. Flash-attention kernel in jax provided for both GPU and TPU. I don't think the problem with GPU is actually a problem with jax. I do know that TPU has some optimization but it still covers mostly in GPU.

@dathudeptrai 
TensorRT-LLM still slower than XLA, are you sure? It seems that XLA does not have capabilities such as rebatching and intermediate result cache, which are of great benefit to improving online throughput.

"FP8 training is already support in Jax"......Are you sure? I just asked Jax guys three months ago. Any updated?
https://github.com/google/jax/issues/16673
https://github.com/google/jax/issues/17176

**Comment by dathudeptrai** [comment]
@MoFHeka 

`Besides, XLA "Flash Attention" supporting was contributed by Nvidia cudnn FMHA kernel. They're still working on it.`

It already finished yesterday so we can use it very soon in both TF and Jax. In jax, actually I am looking to the pallas implementation of flash-attention (https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/attention.py). They even introduced `splash attention` for TPU silently (https://github.com/google/jax/commit/598b46aab5e1595669acc4bb1052b905cca68a9e). 

`"FP8 training is already support in Jax"......Are you sure? I just asked Jax guys three months ago. Any updated?`

Yeah I sure about this because I personally tested FP8 training with jax which is 50% faster than BF16/FP16. You can take a look this PR, the change is very simple (https://github.com/google/flax/pull/3398). 


`"TensorRT-LLM still slower than XLA", are you sure? It seems that XLA does not have capabilities such as rebatching(at batch dim or sequence dim) and intermediate result cache, which are of great benefit to improving online throughput. Also it packages a lot of tools such as parameter quantization, saving us development costs.`

Yeah XLA is faster in my case. I don't know what you mean by rebatching, is that mean dynamic batch size in the inference ?. Basically we can just let the XLA warmed up with the shape you want and you can save the XLA-profile to cache file and then you can load it and use it for later (exactly the same as tensorrt).

**Comment by dathudeptrai** [comment]
@MoFHeka basically I think complier is the future of AI and I pretty much believe in MLIR-based compiler (XLA, Iree). The pytorch side always have to adjust the code to adapt with new training techniques/library, ... and I already felt tired :D. I would rather convert the whole pytorch model to TF or jax for just inferencing with Tensorflow-Tensorrt or XLA :D. XLA can only get better and faster over the time and we don't need to adjust the code to make it faster like pytorch.

**Comment by MoFHeka** [comment]
@dathudeptrai Yes, I believe Google should have practices for various acceleration methods. But I can't seem to get the Google people out to respond. 😂 @jurahul @hawkinsp

Besides, have you ever try zero3 with Jax? When I read the comment in the code, it seems that there're some bugs.

**Comment by MoFHeka** [comment]
@dathudeptrai 
Zero stage 2 is also fine. Does it stable when applied in Jax LLM training? How is the performance?

**Comment by bchetioui** [comment]
@MoFHeka this seems firmly in the real of modelling to me, I don't think it is at all actionable in XLA. The repository for the relevant library you intend to use is probably a better place to file this bug. Some notes:

* [MaxText at least mentions sequence parallelism here](https://github.com/google/maxtext/blob/main/MaxText/max_utils.py#L169-L174). I'm not sure whether that is enough to solve your issue;
* If the offered sharding strategies do not fit your needs, maybe you can also consider trying [shard_map](https://jax.readthedocs.io/en/latest/jep/14273-shard-map.html).

**Comment by MoFHeka** [comment]
@dathudeptrai Saddly, PaxML is much slower than Megatron(TP style) or Deepspeed(FSDP style). I tested LLAMA2_7B and  LLAMA2_13B with single A100 node. And I have already used the latest Nvidia image ghcr.io/nvidia/upstream-pax:nightly-2024-01-26.

**Comment by dathudeptrai** [comment]
@MoFHeka I have not tested the training yet, but did you use bf16 or FP8, and did you use jax.jit correctly ?. Did you used this code for the best performance (https://github.com/NVIDIA/JAX-Toolbox/tree/main/rosetta/rosetta/projects/pax). It is hard to belive that jax is much slower. I am switching to jax more and more modules as I found it is much faster than pytorch in the inference now.

**Comment by MoFHeka** [comment]
@dathudeptrai Of course, I even add more XLA optimization option, but they didn't help....
`--xla_gpu_graph_level=3 --xla_gpu_enable_async_all_reduce=true
 --xla_gpu_enable_async_collectives=true --xla_gpu_enable_async_collective_permute=true
 --xla_gpu_enable_async_all_gather=true --xla_gpu_enable_async_reduce_scatter=true
 --xla_gpu_enable_async_all_to_all=true --xla_gpu_all_reduce_contiguous=true
 --xla_gpu_all_reduce_blueconnect_num_devices_per_host=true
 --xla_gpu_enable_cudnn_frontend=true --xla_gpu_enable_cudnn_fmha=true --xla_gpu_fused_attention_use_cudnn_rng=true
 --xla_gpu_enable_cudnn_layer_norm`

**Comment by MoFHeka** [comment]
@dathudeptrai 
According https://github.com/NVIDIA/JAX-Toolbox/tree/main/rosetta/rosetta/projects/pax, Nvidia train a 5B GPT model with TE BF16 in 256 A100 GPU. And its performance 465.45 Sequences/Sec when sequences global batch size is 8*256=2048. So it means it costed 4.4s per step. Am I correct?
This script could calculate its MFU which is 38.958427%.
```python
# Nvidia Jax GPT5B
card_num=256
gbs=8*card_num
layers=24
num_query=32
num_heads=32
enc_seq_len=2048
hs=4096
ffn_hs=16384
vocab=50304

sequences_per_sec=465.45
seconds_per_step=gbs/sequences_per_sec


#Model total parameters：
params_qkv_state = (1+2*(num_query/num_heads))*hs*hs
params_post_attention_linear = hs*hs
params_fead_forward_network = 2*hs*ffn_hs
params_vocabulary_embedding = hs*vocab


#FPROP：
qkv_state = gbs*2*(1+2*(num_query/num_heads))*enc_seq_len*hs*hs
attention_matrix_computation = gbs*2*enc_seq_len*enc_seq_len*hs
attention_over_values = gbs*2*enc_seq_len*enc_seq_len*hs
post_attention_linear_projection = gbs*2*enc_seq_len*hs*hs
fead_forward_network = gbs*(2*2*enc_seq_len*ffn_hs*hs)
vocabulary_embedding = gbs*2*enc_seq_len*hs*vocab

#BPROP：
#FPROP*2

model_params = (params_qkv_state+params_post_attention_linear+params_fead_forward_network)*layers + params_vocabulary_embedding 
model_float = 3*((qkv_state+attention_matrix_computation+attention_over_values+post_attention_linear_projection+fead_forward_network)*layers + vocabulary_embedding) 
model_flops = model_float/seconds_per_step
cluster_ideal_flops = 312*(10**12) * card_num
MFU = model_flops/cluster_ideal_flops
print("Model parameters {:4f}B MFU={:4f}%".format(model_params/(10**9),MFU*100))
```
